### PR TITLE
 [Syntax] Verify syntax tree only if we reached the real EOF

### PR DESCRIPTION
**Pull request by rintaro**
We used to verify the tree for every `parseIntoSourceFile()` invocation. That used to cause significant performance penalty if `-verify-syntax-tree` is passed and the source contains `TopLevelCodeDecl`s.

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: test/Syntax/diagnostics_verify.swift [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  // expected-error@-1 2 {{unknown expression syntax exists in the source}}
+  // expected-error@-2 {{unknown statement syntax exists in the source}}
+} // expected-error {{expected expression}} expected-error {{unknown statement syntax exists in the source}}
+
+if false {
+  [.]
+  // expected-error@-1 {{expected identifier after '.' expression}}
+  // expected-error@-2 2 {{unknown expression syntax exists in the source}}
+  // expected-error@-3 {{unknown statement syntax exists in the source}}
+} // expected-error {{unknown statement syntax exists in the source}}
```
Comment (rintaro) [inline test/Syntax/diagnostics_verify.swift]: We are still making a lot of unnecessary `Unknown` nodes for invalid source.

Comment (nkcsgexi) [reply test/Syntax/diagnostics_verify.swift]: Can we design some rules to remove unnecessary nodes as post-processing? e.g. removing empty `CodeBlockItem`.

File: lib/Parse/SyntaxParsingContext.cpp [hunk 2]
```diff
@@ -297,7 +297,7 @@ void finalizeSourceFile(RootContextData &RootData,
   assert(newRaw);
   SF.setSyntaxRoot(make<SourceFileSyntax>(newRaw));
 
-  if (SF.getASTContext().LangOpts.VerifySyntaxTree) {
+  if (EOFToken->isPresent() && SF.getASTContext().LangOpts.VerifySyntaxTree) {
```
Comment (nkcsgexi) [inline lib/Parse/SyntaxParsingContext.cpp]: Nice catch! Can we add some comments here why this is checked? If `finalizeSourceFile` gets called before we really finalize it, we may need to rename this function to a more descriptive one.

**Review by nkcsgexi — APPROVED** [review]

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
I wasn't able to come up with the good new name for `finalizeSourceFile`.
@nkcsgexi Any suggestion?

**Pull request merged**